crawl-ref/source/dungeon.cc

```diff
@@ -6129,6 +6130,132 @@ static bool _valid_item_for_shop(int item_index, shop_type shop_type_,
     return true;
 }
 
+/**
+ * Is this item reasonably worthless after the early/mid game? Shops get an
+ * increasing chance to reroll such items the further down we go.
+ */
+static bool _item_is_trash(item_def& item)
+{
+    // Never trash! (Maybe we could apply *some* logic to random artefacts, but
+    // it's really hard to say what may or may not be useful to a given player.
+    if (is_artefact(item))
+        return false;
+    switch (item.base_type)
+    {
+    case OBJ_SCROLLS:
+        switch (item.sub_type)
+        {
+        case SCR_NOISE:
+        case SCR_IDENTIFY:
+            return true;
+        default:
+            return false;
+        }
+    case OBJ_POTIONS:
+        switch (item.sub_type)
+        {
+        case POT_MOONSHINE:
+        case POT_LIGNIFY:
+            return true;
+        default:
+            return false;
+        }
+    case OBJ_WANDS:
+        switch (item.sub_type)
+        {
+        case WAND_FLAME:
+        case WAND_DIGGING: // Questionable
+            return true;
+        default:
+            return false;
+        }
+    // Get a bit more choosy. A lot of base jewellery types are boring once
+    // you've seen one, but ones with variable plusses can still be good.
+    case OBJ_JEWELLERY:
+        switch (item.sub_type)
+        {
+        case RING_SLAYING:
+        case AMU_FAITH:
+        case AMU_REGENERATION:
+        case AMU_MANA_REGENERATION:
+            return false;
+        case RING_STRENGTH:
+        case RING_INTELLIGENCE:
+        case RING_DEXTERITY:
+        case RING_PROTECTION:
+        case RING_EVASION:
+        case RING_MAGICAL_POWER:
+            return item.plus < 5;
+        default:
```

Comment:
I don't really understand what this is trying to achieve. Str/Int/Dex rings already only spawn at +6, protection at +4 and evasion at +5, so at present this bans rings of protection and keeps everything else.

Reply:
This was mistaken as I though these rings had variable plusses generally but obviously that's only artifacts. But see my reply to your other comment on rings below.